app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/reliefdetails/ReliefDetailsController.scala

```diff
@@ -205,51 +214,144 @@ class ReliefDetailsController @Inject() (
     withFillingOutReturnAndReliefDetailsAnswers(request) {
-      case (_, _, _, answers) =>
-        commonDisplayBehaviour(answers)(
-          form = _.fold(
-            _.lettingsRelief.fold(lettingsReliefForm)(a => lettingsReliefForm.fill(a.inPounds())),
-            c => lettingsReliefForm.fill(c.lettingsRelief.inPounds())
-          )
-        )(
-          page = lettingsReliefPage(_, _)
-        )(
-          requiredPreviousAnswer = _.fold(
-            _.privateResidentsRelief,
-            c => Some(c.privateResidentsRelief)
-          ),
-          routes.ReliefDetailsController.privateResidentsRelief()
-        )
+      lettingsPreBehaviour {
+        lettingsDisplayBehaviour
+      }
     }
   }
 
   def lettingsReliefSubmit(): Action[AnyContent] = authenticatedActionWithSessionData.async { implicit request =>
     withFillingOutReturnAndReliefDetailsAnswers(request) {
-      case (_, fillingOutReturn, draftReturn, answers) =>
-        commonSubmitBehaviour(fillingOutReturn, draftReturn, answers)(
-          form = lettingsReliefForm
-        )(page = lettingsReliefPage(_, _))(
-          requiredPreviousAnswer = _.fold(
-            _.privateResidentsRelief,
-            c => Some(c.privateResidentsRelief)
-          ),
-          redirectToIfNoRequiredPreviousAnswer = routes.ReliefDetailsController.privateResidentsRelief()
-        )(
-          updateAnswers = {
-            case (p, draftReturn) =>
-              draftReturn.copy(
-                reliefDetailsAnswers = Some(
-                  answers.fold(
-                    _.copy(lettingsRelief = Some(AmountInPence.fromPounds(p))),
-                    _.copy(lettingsRelief = AmountInPence.fromPounds(p))
-                  )
-                )
-              )
-          }
-        )
+      lettingsPreBehaviour {
+        lettingsSubmitBehaviour
+      }
     }
   }
 
+  def lettingsDisplayBehaviour(
+    fillingOutReturn: FillingOutReturn,
+    draftReturn: SingleDisposalDraftReturn,
+    answers: ReliefDetailsAnswers,
+    lettingsReliefLimit: AmountInPence
+  )(implicit request: RequestWithSessionData[_]): Future[Result] =
+    commonDisplayBehaviour(answers)(
+      form = _.fold(
+        _.lettingsRelief.fold(
+          lettingsReliefForm(answers, lettingsReliefLimit)
+        )(a => lettingsReliefForm(answers, lettingsReliefLimit).fill(a.inPounds())),
+        c => lettingsReliefForm(answers, lettingsReliefLimit).fill(c.lettingsRelief.inPounds())
+      )
+    )(
+      page = lettingsReliefPage(_, _)
+    )(
+      requiredPreviousAnswer = _.fold(
+        _.privateResidentsRelief,
+        c => Some(c.privateResidentsRelief)
+      ),
+      routes.ReliefDetailsController.privateResidentsRelief()
+    )
+
+  private def lettingsPreBehaviour(
+    lettingsPostBehaviour: (
+      FillingOutReturn,
+      SingleDisposalDraftReturn,
+      ReliefDetailsAnswers,
+      AmountInPence
```

Comment:
a lot of the stuff here is provided by `withFillingOutReturnAndReliefDetailsAnswers` already. You could just reuse that have a small method which allows you to get the private resident relief when you need it:
```scala
def withPrivateResidentRelief(answers: ReliefDetailsAnswers)(f: AmountInPence => Future[Result]): Future[Result] ???
```
You'll end up with a lot less code that way

Reply:
then you're code would look something like this:
```
 withFillingOutReturnAndReliefDetailsAnswers(request) {
    case (_, fillingOutReturn, draftReturn, answers) =>
      withTaxYear(draftReturn){ taxYear =>
         ....
     }
}
```

Reply:
Yes, nice approach, done.